lens/fetcher.go

```diff
@@ -0,0 +1,397 @@
+	doc := mapping.NewDoc()
+
+	for fieldName, fieldByteValue := range docAsMap {
+		if fieldName == request.KeyFieldName {
+			key, ok := fieldByteValue.(string)
+			if !ok {
+				return core.Doc{}, core.ErrInvalidKey
+			}
+
+			doc.SetKey(key)
+			continue
+		}
+
+		fieldDesc, fieldFound := df.fieldDescriptionsByName[fieldName]
+		if !fieldFound {
+			// Note: This can technically happen if a Lens migration returns a field that
+			// we do not know about. In which case we have to skip it.
+			continue
+		}
+
+		fieldValue, err := core.Decode(fieldDesc, fieldByteValue)
+		if err != nil {
+			return core.Doc{}, err
+		}
+
+		index := mapping.FirstIndexOfName(fieldName)
+		doc.Fields[index] = fieldValue
+	}
+
+	if value, ok := docAsMap[request.DeletedFieldName]; ok {
+		if wasDeleted, ok := value.(bool); ok {
+			if wasDeleted {
+				doc.Status = client.Deleted
+			} else {
+				doc.Status = client.Active
+			}
+		}
+	}
+
+	doc.SchemaVersionID = df.col.Schema.VersionID
+
+	return doc, nil
+}
+
+// lensDocToClientDoc converts a LensDoc to a client.Document.
+func (df *lensedFetcher) lensDocToClientDoc(docAsMap LensDoc) (*client.Document, error) {
+	key, err := client.NewDocKeyFromString(docAsMap[request.KeyFieldName].(string))
+	if err != nil {
+		return nil, err
+	}
+	doc := client.NewDocWithKey(key)
+
+	for fieldName, fieldByteValue := range docAsMap {
+		if fieldName == request.KeyFieldName {
+			continue
+		}
+
+		fieldDesc, fieldFound := df.fieldDescriptionsByName[fieldName]
+		if !fieldFound {
+			// Note: This can technically happen if a Lens migration returns a field that
+			// we do not know about. In which case we have to skip it.
+			continue
+		}
+
+		fieldValue, err := core.Decode(fieldDesc, fieldByteValue)
+		if err != nil {
+			return nil, err
+		}
+
+		err = doc.SetAs(fieldDesc.Name, fieldValue, fieldDesc.Typ)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	doc.SchemaVersionID = df.col.Schema.VersionID
+
+	// Note: client.Document does not have a means of flagging as to whether it is
+	// deleted or not, and, currently the fetcher does not ever returned deleted items
+	// from the function that returs this type.
+
+	return doc, nil
+}
+
+// updateDataStore updates the datastore with the migrated values.
+//
+// This removes the need to migrate a document everytime it is fetched as the second time around
+// the underlying fetcher will return the migrated values cached in the datastore instead of the
+// underlying dag store values.
+func (df *lensedFetcher) updateDataStore(ctx context.Context, original map[string]any, migrated map[string]any) error {
+	modifiedFieldValuesByName := map[string]any{}
+	for name, originalValue := range original {
+		migratedValue, ok := migrated[name]
+		if !ok {
+			// If the field is present in the original, and missing from the migrated, it
+			// means that a migration has removed it, and we should set it to nil.
+			modifiedFieldValuesByName[name] = nil
```

Comment:
todo: Add test for this (requires new lens module)

- [ ] Test clearing of field